File: python/gamemenu.py
"Game main menu handler, part of SPACEWAR."

import math, os
import pygame, pygame.draw
from pygame.locals import *
import var
import gfx, snd, txt
import input
import gamecreds, gamenews, gamepref, gameplay


images = []
boximages = []
yboximages = []
rboximages = []
fame = None

class MenuItem:
    def __init__(self, imgname, handler):
        self.imgname = imgname
        self.handler = handler

    def init(self, pos):
        self.img_on = gfx.load('menu_'+self.imgname+'_on.png')
        self.img_off = gfx.load('menu_'+self.imgname+'_off.png')
        self.rect = self.img_on.get_rect().move(pos)
        self.smallrect = self.img_off.get_rect()
        self.smallrect.center = self.rect.center


def load_game_resources():
    global images, boximages, fame
    images = []
    images.append(gfx.load('menu_on_bgd.png'))
    images[0].set_colorkey(0)
    images.append(gfx.load('logo.png'))
    images.append(gfx.load('ship-big.png'))
    images[1].set_colorkey()
    images[2].set_colorkey()

    global boximages, yboximages, rboximages
    imgs = gfx.load_raw('bigboxes.png')
    origpal = imgs.get_palette()
    boximages = gfx.animstrip(imgs)
    pal = [(g,g,b) for (r,g,b) in origpal]
    imgs.set_palette(pal)
    yboximages = gfx.animstrip(imgs)
    pal = [(g,b,b) for (r,g,b) in origpal]
    imgs.set_palette(pal)
    rboximages = gfx.animstrip(imgs)

    fame = gfx.load('fame.png')

    snd.preload('select_move', 'select_choose')



class GameMenu:
    def __init__(self, prevhandler):
        self.prevhandler = prevhandler
        self.current = 0
        self.glow = 0.0
        self.switchhandler = None
        self.switchclock = 0
        self.startclock = 5
        self.logo = images[1]
        self.logorect = self.logo.get_rect().move(100, 25)
        self.logorectsmall = self.logorect.inflate(-2,-2)
        self.boxtick = 0
        self.boximages = boximages
        self.boxrect = self.boximages[0].get_rect().move(580, 80)
        self.bigship = images[2]
        self.bigshiprect = self.bigship.get_rect().move(450, 250)

        if var.ai_train:
            print "ai_train"
            Player = gameplay.AIPlay
        else:
            print "not ai_train"
            Player = gameplay.GamePlay

        self.menu = [
            MenuItem('start', Player),
            MenuItem('news', gamenews.GameNews),
            MenuItem('creds', gamecreds.GameCreds),
            MenuItem('setup', gamepref.GamePref),
            MenuItem('quit', None),
        ]

        pos = [20, 380] #[100, 420]
        odd = 0
        for m in self.menu:
            m.init(pos)
            pos[0] += 150
            odd = (odd+1)%2
            if odd:
                pos[1] += 20
            else:
                pos[1] -= 20


    def starting(self):
        snd.playmusic('aster2_sw.xm')

        gfx.dirty(gfx.surface.blit(self.logo, self.logorect))
        gfx.dirty(gfx.surface.blit(self.bigship, self.bigshiprect))

        fnt = txt.Font(None, 18)
        self.version = fnt.text((200, 120, 100), 
                'Spacewar Version ' + var.version, (10, 580), 'topleft')

        fnt = txt.Font(None, 22)
        if var.ai_train:
            self.players = fnt.text((200, 175, 120), 
                'AI Training', (115, 500), 'center')
        else:
                if var.game_mode == 2:
                    self.players = fnt.text((200, 175, 120), 
                        '1 player tutorial', (115, 500), 'center')
                else:
                    players = var.player_cnt()
                    self.players = fnt.text((200, 175, 120), 
                        'with %d players' % (players), (115, 500), 'center')

    def quit(self):
        self.current = len(self.menu)-1
        self.workbutton()


    def workbutton(self):
        button = self.menu[self.current]
        if not button.handler:
            self.switchhandler = self.prevhandler
        else:
            self.switchhandler = button.handler
        self.switchclock = 10


    def clearitem(self, item, dirty=0):
        r = self.background(item.rect)
        if dirty:
            gfx.dirty(r)


    def setalphas(self, alpha, extras=[]):
        imgs = []
        menu = self.menu
        selected = menu[self.current]
        imgs.extend([x.img_off for x in menu if x is not selected])
        imgs.extend(extras)
        if gfx.surface.get_bytesize()==1:
            return
        if alpha < 255:
            for i in imgs:
                i.set_alpha(alpha)
        else:
            for i in imgs:
                i.set_alpha()
                c = i.get_colorkey()
                if c:
                    i.set_colorkey(c, RLEACCEL)


    def drawitem(self, item, lit):
        if not lit:
            gfx.surface.blit(item.img_off, item.smallrect)
        else:
            lite = images[0]
            glowval = (math.sin(self.glow) + 2.5) * 50.0
            if self.switchclock == 2:
                glowval /= 2
            if gfx.surface.get_bytesize()>1:
                lite.set_alpha(glowval)
            gfx.surface.blit(lite, item.rect)
            gfx.surface.blit(item.img_on, item.rect)
        gfx.dirty(item.rect)


    def input(self, i):
        if i.release:
            return
        if self.switchclock:
            return
        if i.translated == input.LEFT:
            self.current = (self.current - 1)%len(self.menu)
            snd.play('select_move')
        elif i.translated == input.RIGHT:
            self.current = (self.current + 1)%len(self.menu)
            snd.play('select_move')
        elif i.translated == input.PRESS:
            self.workbutton()
            snd.play('select_choose')
        elif i.translated == input.ABORT:
            snd.play('select_choose')
            self.quit()

    def event(self, e):
        pass


    def run(self):
        menu = self.menu
        self.glow += .35
        self.boxtick = (self.boxtick + 1)%15
        boximg = self.boximages[self.boxtick]

        if self.startclock:
            alpha = (6-self.startclock)*40
            self.setalphas(alpha, [menu[self.current].img_on, boximg])
            self.background(self.boxrect)
        elif self.switchclock:
            alpha = (self.switchclock-1)*20
            self.setalphas(alpha, [boximg])
            self.background(self.boxrect)
            if self.switchclock == 2 and gfx.surface.get_bytesize()>1:
                menu[self.current].img_on.set_alpha(128)

        for m in menu:
            self.clearitem(m)

        gfx.updatestars(self.background, gfx)

        gfx.dirty(gfx.surface.blit(*self.version))
        gfx.dirty(gfx.surface.blit(*self.players))

        if self.startclock == 1 or self.switchclock == 1:
            self.setalphas(255, [menu[self.current].img_on] + self.boximages)

        if self.switchclock != 1:
            r = gfx.surface.blit(boximg, self.boxrect)
            gfx.dirty(r)

            select = menu[self.current]
            for m in [m for m in menu if m is not select]:
                self.drawitem(m, 0)
            self.drawitem(select, 1)

        else:
            for m in menu:
                gfx.dirty(m.rect)
            if self.switchhandler == self.prevhandler:
                var.handler = self.prevhandler
            else:
                var.handler = self.switchhandler(self)
                self.switchclock = 0
                self.switchhandler = None
                self.startclock = 5
            gfx.dirty(self.background(self.version[1]))
            gfx.dirty(self.background(self.players[1]))
            gfx.dirty(gfx.surface.fill((0, 0, 0), self.logorect))
            gfx.dirty(gfx.surface.fill((0, 0, 0), self.bigshiprect))

        if self.startclock:
            self.startclock -= 1
        elif self.switchclock:
            self.switchclock -= 1



    def background(self, area):
        fullr = gfx.surface.fill((0, 0, 0), area)
        if self.switchclock != 1:
            if area.colliderect(self.bigshiprect):
                    r = area.move(-self.bigshiprect.left, -self.bigshiprect.top)
                    return gfx.surface.blit(self.bigship, area, r)
            elif self.switchclock != 1:
                if area.colliderect(self.logorectsmall):
                    r = area.move(-self.logorect.left, -self.logorect.top)
                    return gfx.surface.blit(self.logo, area, r)
        return fullr

File: js/gamemenu.js
"use strict";

var MenuItem = function (game, name, state) {
    Phaser.Sprite.call(this, game, 0, 0, 'menu-'+name+'-off')
    this.visible = true
    this.anchor.set(0.5, 0.5)

    this.bgd_sprite = game.add.sprite(0, 0, 'menu-bgd')
    this.bgd_sprite.visible = false
    this.bgd_sprite.anchor.set(0.5, 0.5)

    this.on_sprite = game.add.sprite(0, 0, 'menu-'+name+'-on')
    this.on_sprite.visible = false
    this.on_sprite.anchor.set(0.5, 0.5)

    this.state = state
    this.glow = 0.0
    this.active = false
}
MenuItem.prototype = Object.create(Phaser.Sprite.prototype)
MenuItem.prototype.constructor = Mass

MenuItem.prototype.init = function(x, y) {
    this.x = x
    this.y = y
    this.bgd_sprite.x = x
    this.bgd_sprite.y = y
    this.on_sprite.x = x
    this.on_sprite.y = y
}

MenuItem.prototype.on = function(x, y) {
    this.visible = false
    this.active = true
    this.bgd_sprite.visible = true
    this.on_sprite.visible = true
}

MenuItem.prototype.off = function(x, y) {
    this.visible = true
    this.active = false
    this.bgd_sprite.visible = false
    this.on_sprite.visible = false
}

MenuItem.prototype.update = function() {
    if (this.bgd_sprite.visible) {
        this.glow += 0.1
        this.bgd_sprite.alpha = Math.abs(Math.sin(this.glow))
    }
}

var GameMenu = function(game) {
    this.current = 0
}

GameMenu.prototype.init = function () {
}

GameMenu.prototype.create = function () {
    var self = this,
        game = this.game

    // Add and start the menu soundtrack
    if (!game.music.menu.isPlaying) {
        game.music.menu.play()
    }

    //'menu-bgd'

    // Define star group
    game.groups = {stars:      starGroup(game),
                   background: game.add.group(),
                   menu:       game.add.group()}

    // "Spacewar" Logo
    this.logo = game.groups.background.create(100, 25, 'logo')

    // Large ship image
    this.ship = game.groups.background.create(450, 250, 'ship-big')
    this.ship.tint = 0x808080

    // Rotating box image/animation
    this.bigboxes = game.groups.background.create(580,  80, 'bigboxes')
    this.bigboxes.animations.add('anim', [0,1,2,3,4,5,6,7,8,9,
                                          10,11,12,13,14], 30, true)
    this.bigboxes.animations.play('anim')
    this.bigboxes.tint = 0xe0e0e0

    this.info = game.add.text(20, 570,
            'Spacewar Version ' + VERSION,
            {font: "Arial Black",
             fontSize: 10,
             fill: rgba(200, 120, 100)},
            game.groups.menu)
    this.info.setShadow(2, 2, 'rbga(128,128,128,0.7)', 2)

    this.version = game.add.text(100, 460,
            'with ' + vars.player_cnt() + ' players',
            {font: "Arial Black",
             fontSize: 11,
             fill: rgba(200, 175, 120)},
            game.groups.menu)
    this.version.setShadow(2, 2, 'rbga(128,128,128,0.7)', 2)

    // Menu items
    this.menu = [new MenuItem(game, 'start', 'Play'),
                 new MenuItem(game, 'setup', 'Setup'),
                 new MenuItem(game, 'news',  'News'),
                 new MenuItem(game, 'creds', 'Creds')]

    var posX = 140, posY = 400
    for (var i in this.menu) {
        var m = this.menu[i]
        game.groups.menu.add(m)
        m.init(posX, posY)
        posX += 180
        if (i % 2) {
            posY -= 30
        } else {
            posY += 30
        }
    }

    // Activate starting menu item
    this.menu[this.current].on()

    //
    // Menu interactions
    //
    function move(dx, x) {
        var len = self.menu.length
        self.menu[self.current].off()
        if (typeof x === 'number') {
            self.current = x
        } else {
            self.current = (self.current + len + dx) % len
        }
        self.menu[self.current].on()
        game.sounds.menu_move.play()
    }
    function choose() {
        if (self.current === 0) {
            game.music.menu.stop()
        }
        game.sounds.menu_choose.play()
        self.state.start(self.menu[self.current].state);
    }

    // Add keyboard interaction
    var k = game.input.keyboard
    k.addKey(Phaser.Keyboard.LEFT ).onDown.add(function() { move(-1) })
    k.addKey(Phaser.Keyboard.RIGHT).onDown.add(function() { move(+1) })
    k.addKey(Phaser.Keyboard.ENTER   ).onDown.add(choose)
    k.addKey(Phaser.Keyboard.SPACEBAR).onDown.add(choose)
    game.input.onDown.add(choose)

    // Add mouse interaction
    game.input.addMoveCallback(function(pointer, absX, absY) {
        var x = absX/game.scaleVal, y = absY/game.scaleVal
        for (var i in self.menu) {
            var menu = self.menu[i]
            if (menu.active) { continue }
            if (Phaser.Rectangle.contains(menu.getBounds(), x, y)) {
                move(null, parseInt(i))
            }
        }
    }, this)
}

GameMenu.prototype.update = function () {
}

GameMenu.prototype.render = function () {
    var game = this.game
    //game.debug.soundInfo(game.music.menu, 20, 460)
}
